### ISSUE TITLE
[no-inferrable-types] Equivalent for [explicit-function-return-types]

### ISSUE THREAD

**Issue by vegerot** [typescript-eslint/typescript-eslint#2673]
tl;dr I would like a rule to ban inferable function return types, similar to how `no-inferrable-types` acts on variables.

- [x ] I have tried restarting my IDE and the issue persists.
- [x ] I have updated to the latest version of the packages.
- [ x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

```JSON
{
  "rules": {
    "@typescript-eslint/no-inferrable-types": ["error"],
    "@typescript-eslint/explicit-function-return-types": ["error"],
  }
}
```

```TS
//err
const a: number = 1
const b = 1

function c(): number {return 1}
//err
function d() {return 1}
```

**Expected Result**

If I had

```JSON
{
  "rules": {
    "@typescript-eslint/no-inferrable-types": ["error"],
    "@typescript-eslint/no-inferrable-function-types": ["error"],
  }
}
```

I would expect all inferable types to be banned

```TS
//err
const a: number = 1
const b = 1

//err
function c(): number {return 1}
function d() {return 1}
```

**Actual Result**
Instead, since that second rule that doesn't exist, I'm stuck with 

```TS
//err
const a: number = 1
const b = 1

function c(): number {return 1}
//err
function d() {return 1}
```
**Additional Info**

I would like a rule to ban inferable function return types, similar to how `no-inferrable-types` acts on variables.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `latest` |
| `@typescript-eslint/parser`        | `latest` |
| `TypeScript`                       | `latest` |
| `ESLint`                           | `latest` |
| `node`                             | `latest` |

**Comment by jimmywarting** [comment]
👍

**Comment by Morozzko** [comment]
someone have this settings work?

**Comment by armano2** [comment]
instead of `@typescript-eslint/no-inferrable-function-types` you should try `@typescript-eslint/no-inferrable-function-type`

[playground](https://typescript-eslint.io/play/#ts=5.2.0-beta&fileType=.tsx&code=PTCmCdwWAKAYwPYDsDOAXABAQwFwaQK4C2ARhBgLwYCMsiqmJlNssIE0MAZgUnGgEtkGOAAoAlHkKlyAbwzhQaAuCQ0MAX1g8%2Bg4QBMJGeYuWr1GoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDNFpoAQwBGSWkVKoMAbXDYc-aB2iQ5EALoAaNVAIlyVGvWSMWrRAA9ijMpTo9YTMvkocmtaInyxo7iYnQwWXkFAWVVeXU5AF8QaKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by vegerot** [comment]
I opened this issue when I was a naive child.  I now see the error of my ways.

**Comment by rentalhost** [comment]
@vegerot I believe it is still a valid addition. Why do you think this would be a bad idea now? I know there are cases where the type needs to be explicit for the return to be correctly 'filtered.' But there are straightforward cases, like the ones you mentioned, that would be great. Less clutter in the code for obvious things.

**Comment by norflin321** [comment]
> I opened this issue when I was a naive child. I now see the error of my ways.

so when you again realize that you were a naive child leaving this comment 😄

you can do it like this
```JSON
"rules": {
  "no-restricted-syntax": [
    "error",
    {
      "selector": "FunctionExpression[returnType.type='TSTypeAnnotation']",
      "message": "function return type trivially inferred from a function definition, remove type annotation"
    }
  ]
}
```
@Morozzko @rentalhost

**Comment by bradzacher** [comment]
@norflin321 that is not a solution to this problem at all and instead just bans return type annotations completely.

**Comment by norflin321** [comment]
> @norflin321 that is not a solution to this problem at all and instead just bans return type annotations completely.

as far as i know typescript always can infer correct return type if your identifiers are types inside function body

**Comment by vegerot** [comment]
I opened this issue when I was young and dumb.  I would never use such a rule

**Comment by bradzacher** [comment]
> as far as i know typescript always (or 99% of the time) can infer correct return type if your identifiers are types inside function body

```ts
function foo() {
    return { prop: 'foo' };
}
```

The return type of this function is `{ prop: string }`.
But you forgot the return type and so the return type should have been `MyType` which is `type MyType = { prop: 'foo' }`.
Now you can't use your function because its return value is incorrect and your lint rule prevents you from annotating the return type. So you have to hack around the lint rule and extract the object to a variable and annotate the variable so that you can indirectly annotate your return type. Banning all return type annotations just creates worse code.

[playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAsiAq5oF4oG8pgE4HswC4oByAMxxyKgF8AoGkgVwDsBjYASxyajJwAoAlOhpRRUAPTixUAHoB+EWKwRgDLNwzY8hUuUpUA3DVr1mbTtwBGAQyyDCcRJGHTJ0+YtHLV69JlwExLz6Ria8glAAztYckSTsEJGwCEgGElKAMuQ0NnZC0bHxiclOEGlugKDkNEA).

### ***Banning all return type annotations is an anti-pattern***

**Comment by Morozzko** [comment]
> > 
> > ### _**Banning all return type annotations is an anti-pattern**_

it's you own mention, i just ask how to realise it

**Comment by norflin321** [comment]
> > as far as i know typescript always (or 99% of the time) can infer correct return type if your identifiers are types inside function body
> 
> ```ts
> function foo() {
>     return { prop: 'foo' };
> }
> ```
> 
> The return type of this function is `{ prop: string }`. But you forgot the return type and so the return type should have been `MyType` which is `type MyType = { prop: 'foo' }`. Now you can't use your function because its return value is incorrect and your lint rule prevents you from annotating the return type. So you have to hack around the lint rule and extract the object to a variable and annotate the variable so that you can indirectly annotate your return type. Banning all return type annotations just creates worse code.
> 
> [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAsiAq5oF4oG8pgE4HswC4oByAMxxyKgF8AoGkgVwDsBjYASxyajJwAoAlOhpRRUAPTixUAHoB+EWKwRgDLNwzY8hUuUpUA3DVr1mbTtwBGAQyyDCcRJGHTJ0+YtHLV69JlwExLz6Ria8glAAztYckSTsEJGwCEgGElKAMuQ0NnZC0bHxiclOEGlugKDkNEA).
> 
> ### _**Banning all return type annotations is an anti-pattern**_

the solution for this issue is easy you can create a variable inside a function and explicitly specify its type, then return this variable from the function:
```TypeScript
type MyType = { prop: "foo" };
function foo() {
  const out: MyType = { prop: "foo" };
  return out;
}
foo() satisfies MyType; // correct
```
or you can make i shorter and use `as` keyword:
```TypeScript
type MyType = { prop: "foo" };
function foo() {
  return { prop: "foo" } as MyType;
}
foo() satisfies MyType; // correct
```

**Comment by bradzacher** [comment]
> the solution for this issue is easy you can create a variable inside a function and explicitly specify its type, then return this variable from the function

Correct, I mentioned this in my comment:

> So you have to hack around the lint rule and extract the object to a variable and annotate the variable so that you can indirectly annotate your return type.

---

> or you can make it shorter using as keyword:

You could - but on doing so you just introduced a safety hole into your codebase because as assertions are unsafe:

```ts
type MyType = { prop: "foo", propWeJustMissed: number };
function foo() {
  // no ts error!!
  return { prop: "foo" } as MyType;
}
foo().propWeJustMissed.toString()
// oops runtime crash :(
```

I'll close my responses with the same sentiment from my last comment.

> Banning all return type annotations just creates worse code.

**Comment by kirkwaiblinger** [comment]
@anasbud 

The trouble is... TS will basically always come up with a return type, no matter what you return. It's just the union of whatever you return. For example,
```ts
function foo() {
   if (Math.random() < 0.5) {
      return 'string';
   } else {
      return 42;
   }
}

foo;
//^? function foo(): "string" | 42
```

So, "prevent you from explicitly writing what Typescript already knows" _implies_ no explicitly specified return types, ever.

Whether or not one prefers or dislikes inferred return types (disclaimer - I prefer them always to be explicit), one would need to come up with some understandable criteria that ban some but not all return type annotations, in order for this issue to be actionable. No such criteria have been proposed in this thread yet.

Perhaps someone would like to have a stab at it? If not, I think we should close this issue.

**Comment by Josh-Cena** [comment]
@kirkwaiblinger That means the rule would error if you write `: "string" | 42` as the return type, and no error if you write `: string | number`, or `type Answer = "string" | 42` and `: Answer`.

**Comment by kirkwaiblinger** [comment]
@Josh-Cena 

I see.... so you're thinking "If the return type is the _same_ as what would have been inferred, ban it. Otherwise allowed."?

**Comment by JoshuaKGoldberg** [comment]
Coming over from https://github.com/typescript-eslint/typescript-eslint/issues/9764#issuecomment-2278956933: now that we can use `isTypeAssignableTo` (#7936), is there any reason to limit this to _same_? Is there any reason not to expand it to _same or wider_?

```ts
// I would hope this would error?
function giveText(): string {
  return 'abc';
}
```

**Comment by Josh-Cena** [comment]
No because it would be intended.

```ts
function mayReturnStringOrNumber(): string | number {
  return "Currently is just a constant: will return number later";
}
```

Here, if the rule forces you to remove the type, then later the addition of `number` would be a breaking change.

I don't think we can even use "X is assignable to Y and Y is assignable to X" without shooting down a lot of valid code. For exmaple, `arr.map((x): SomeComplexObjectType => ({ ... }))` is arguably much better since you get a nicely aliased `SomeComplexObjectType[]` return type instead of a huge object literal type.

**Comment by bradzacher** [comment]
Honestly I'm a -1 on this rule because it's so vague.
I don't think there's any world where you can define the rule such that it would be truly predictable and work without surprises.

For example
```ts
interface T { a: string }
function foo(arg: string): T {
  return { a: arg };
}
```

Should this report?
Technically yes because the object type `{ a: string }` is the same as the type `T`.
But also no it shouldn't because `{ a: string }` is an anonymous type and `T` is a named type -- so they're also not the same thing.

This applies to things like
```ts
type Path = string
function asPath(arg: string): Path {
  if (someComplicatedLogicIsFalse()) {
    return coerceArgToPath(arg);
  }
  return arg;
}
```

Should the rule report here? Well maybe in future we wanted to make `Path` a branded type but the rule forced us to remove annotations.

I just think there are so many cases that are ambiguous and you would end up with surprising results.

Additionally in FOUR YEARs this request has garnered just 8 unique reactions.
There's just not a community want for this rule.
The OP has even stated that since they filed this they've changed their mind and wouldn't use this rule.

So yeah, big -1 from me.

**Comment by kirkwaiblinger** [comment]
Personally, my thoughts here align with @bradzacher's

**Comment by bradzacher** [comment]
@Josh-Cena is a +1, so that leaves @JoshuaKGoldberg.

**Comment by bradzacher** [comment]
In this function, `typeof (typeof v === 'boolean') === boolean` and the return type is `boolean`
```ts
function isBoolean(v: unknown): boolean {
  return typeof v === 'boolean';
}
```
But if you remove the return type now the return type is inferred as `v is boolean`, changing the return type and potentially causing downstream type errors.

See specialisation of this rule in #9764

---

```ts
declare function querySelector<T>(selector: string): T;
interface MyNode { type: 'node' }

function myFunction1(): MyNode {
  return querySelector('node');
}
function myFunction2() {
  //       ^?
  return querySelector('node');
}
```
[playground](https://www.typescriptlang.org/play/?ssl=2&ssc=34&pln=1&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwEdkQYBPAZRAnAxxgB4AVAPgAoBnKmugLnnYwwsqAOYBKPowDcAKGEYSiKGAQBZUgDkcoeAG94GUgAcQfAOSptIM-AC+MmSnTY88ALakAYmky5UARlYJeHUtHV0ZeHg4DGQYfCISCi5MOlYLKzMxWXsnX1cPb2c-ACYgvUj4AHoqqLqogD0AfkqYuITiMkpqVJh0y1AsnKA)
In this example:
`myFunction1`'s explicit return type is `MyNode` and `typeof querySelector('node')` is inferred to be `MyNode` -- so the rule will report. However as illustrated by `myFunction2` -- without the annotation the return type is instead inferred to be `unknown`.

This is a surprising result that we've similarly seen from `no-unnecessary-assertion`.

---

```ts
function myFunction1(): string {
  return 'a';
}
function myFunction2(): 'a' {
  return 'a';
}
function myFunction3() {
  //     ^?
  return 'a';
}
```
[playground](https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAWwJ4DFzXmAjACgEoAuRAZygCcYwBzRAbwChFFKBTKESpAcgENeAbiYBfJqEiwEKDFmlgATEVIDejFm07c+gkeMnYZaTFJwBmIhtYB6G6wcA9APyaOXHojX6mQA)

Which annotation is "the most correct"? Both are correct and technically by definition both are inferable.

---

My point:
There is large complexity to the rule and all the edge cases.
It's definitely possible for us to spec something out and we could likely get something reasonably predictable given enough time and documentation.

But in 4 years we haven't even had TEN community interactions with this issue.
So it's not something the community is interested in using.

Hence my conclusion is that the value is much, much lower than the very high cost of implementation and what will likely be a high cost of maintenance (at least in the short term as we iron out the documentation and edge cases).

**Comment by Josh-Cena** [comment]
> But if you remove the return type now the return type is inferred as v is boolean, changing the return type and potentially causing downstream type errors.

Before inferred predicates, it would error if you annotated `: boolean` and not error if you annotated `: v is boolean`. Afterwards, it would be the other way. (As I said above, it would error as long as the types are not nominally equivalent.) This is strictly better than a rule that always enforces type annotations because it surfaces any changes in the inferrable types.

> This is a surprising result that we've similarly seen from no-unnecessary-assertion

I agree it's surprising but we already have similar behavior. I don't think we should not build a rule just because of certain limitations that users are already aware of.

> Which annotation is "the most correct"? Both are correct and technically by definition both are inferable.

That depends on what TS thinks? If TS infers `'a'` then we report `: string` and vice versa.

**Comment by bradzacher** [comment]
> That depends on what TS thinks? If TS infers 'a' then we report : string and vice versa. 

Well funny story cos the answer is both! Or rather "either depending on the context of the return value".

```ts
declare const x: "a";
function f() {
    //   ^? () => "a"
    return x;
}
function g() {
    //   ^? () => string
    return "a";
}
```

This surprising nature is the sort of logic that we would need to encode into the rule. Note that it is not evidently clear from the type objects alone that this is the case and instead we would need to do syntactic checks.

This is the sort of case that can seem surprising and buggy to users.

> I agree it's surprising but we already have similar behavior.

Similar, buggy behaviour that we just haven't gotten around to fixing yet. A new rule would be required to not have the same bug given it's well known, of course.

---

My comment was just intended to be some examples of hairy cases. There will be many more that would be uncovered through implementation and more still through initial user testing.

I just don't think the user interest in such a rule existing (a small percentage going by the community interactions) is large enough to outweigh the likely large cost of design, implementation and support for such a rule.

Either way - I've said my piece. I'm a strong -1 and there isn't anything that is going to convince me otherwise barring a overwhelming outpouring of support from the community in favour of it.

**Comment by Josh-Cena** [comment]
If it's significantly more difficult than "ask TS what would the inferred type be if there's no type annotation" then I also don't think there's significant ROI. I would love to be a user of this rule though.

**Comment by bradzacher** [comment]
> If it's significantly more difficult than "ask TS what would the inferred type be if there's no type annotation"

This is not possible. TS cannot do speculative querying. We can only make educated guesses based on the rules we know exist.

So that means we must query the functions control flow graph to collect all its termination points so that we can guess at what TS will use as the return type for each one, then we can manually recreate the resulting union type and compare it against the annotated return type.

Which is a whole thing that I haven't even gone into or even done before in a rule -- control flow analysis is complicated. All the examples we've been looking at are simple zero branch, single return functions.

But like here's the next level of complexity
```ts
function foo(b: boolean): string | number {
  if (b) {
    return 1;
  }
  return 'a';
}
```

Then there's `throw` and `never` returning function calls, assertion functions to consider.

At a high level this rule would be:
1) if a function has an explicit return type annotation
1) using its control flow graph, collect all of its valid return points
1) for each return point guess the calculated type for the return and construct a union type by hand
1) compare this with the real annotation and determine, based on our "pre-defined and predictable rules", whether or not the annotation is "bad"

Step 3 will be a doooooozy

**Comment by JoshuaKGoldberg** [comment]
...at this point I'm fine with dropping the issue. I still would _really really like_ to have the rule exist, but Brad's points on the difficulties of it have scared me away from wanting to tackle it in typescript-eslint. 😣

**Comment by Josh-Cena** [comment]
^Exactly same thought.

**Comment by JoshuaKGoldberg** [comment]
We have consensus.

To anybody reading this and feeling 👎 on the decision: we're not denying that the rule can exist! Just that we, as a community open source project, have nowhere near enough time to maintain it. You're welcome to use the docs in https://typescript-eslint.io/developers/custom-rules to build the rule yourself.

In fact, I'd request someone _please_ build this rule & prove us wrong. It'd be a great one to have exist. 😄 

Thanks all!
